registry/data-models/models.py

```diff
@@ -0,0 +1,142 @@
+                             | 
+                +------------------------------+    
+                |                              |      
+                | 1:n                          | 1:n         
+       +------------------+           +------------------+
+       | FeatureInterface |           |     Anchor       |
+       +------------------+           +------------------+
+               ｜                             ｜        |
+               ｜ 1:n                         ｜1:n     |
+               +------------------------------+        |
+                             ｜                         |
+                             ｜1:n                      | contains
+                      +---------------+           +---------------+
+                      |    Feature    | --------- |    Source     |
+                      +---------------+ contains  +---------------+
+                              |
+                              |
+                      +----------------+
+                      | Transformation |
+                      +----------------+
+"""
+
+
+class FeatureValue(BaseModel):
+    """
+    A value of a Feature.
+    NOTE: FeatureValue is to be customized in each registry
+    """
+    pass
+
+
+class FeatureValueType(BaseModel):
+    """
+    A value of a Feature.
+    NOTE: FeatureValue is to be customized in each registry
+    """
+    pass
+
+
+class Source(BaseModel):
+    """
+    Type of FeatureValue.
+    It defines where the feature is extracted from. It can be online, offline, Restli, or other FeatureInterfaces.
+    """
+    pass
+
+
+class DataSource(Source):
+    pass
+
+
+class FeatureSource(Source):
+    input_feature_interface_ids: List[str]  # List of input feature interface IDs
+    pass
+
+
+class Transformation(BaseModel):
+    """
+    The transformation of a Feature.
+    A transformation function represents the transformation logic to produce feature value from the source of FeatureAnchor
+    NOTE: Transformation is to be customized in each registry
+    """
+    pass
+
+
+class Feature(BaseModel):
+    """
+    Actual implementation of FeatureInterface.
+    An implementation defines where a feature is extracted from (Source) and how it is computed (Transformation).
+    The Source of a feature can be raw data sources (like Rest.li, HDFS) and/or other features.
+    """
+    feature_interface_id: str  # ID of the feature interface that the feature belongs to
+    id: str
+    name: str
+    source: Source  # Source can be either data source or feature source
+    transformation: Transformation  # transformation logic to produce feature value
+
+
+class AnchorFeature(Feature):
+    """
+    Feature implementation of FeatureInterface which anchored to a data source.
+    """
+    source: DataSource
+
+
+class DerivedFeature(Feature):
+    """
+    Feature implementation that is derived from other FeatureInterfaces.
+    """
+    source: FeatureSource
+
+
+class FeatureInterface(BaseModel):
```

Comment:
I think we should call this "FeatureName" instead. Name is the semantic identifier we use to identify a feature. I think "Name" works, no matter if we are using string or URN identifier because a URN is a type of name anyways (Uniform Resource Name)

Reply:
Sure.